components/donateTicketForm.js

```diff
@@ -0,0 +1,231 @@
+      } else if (frequency === fundraisingEventTicketPatron) {
+        setFieldValue('amount', '200')
+      }
+    }
+    handleChange(e)
+  }
+
+  donate = async (formValues) => {
+    this.setLocalState({ loading: true })
+    let token
+    try {
+      const cardElement = this.props.elements.getElement(CardElement)
+      const res = await this.props.stripe.createToken(cardElement)
+      token = res.token
+    } catch (e) {
+      this.setState({ error: e.message, loading: false })
+      return
+    }
+
+    if (!token) {
+      this.setState({ error: 'Invalid CC info!', loading: false })
+      return
+    }
+    try {
+      const { frequency, firstName, lastName, email, amount } = formValues
+      const responseStream = await fetch('/api/purchase-ticket', {
+        method: 'POST',
+        body: JSON.stringify({
+          source: token,
+          firstName,
+          frequency,
+          lastName,
+          amount: amount * 100,
+          email
+        })
+      })
+      const response = await responseStream.json()
+      if (response.success) {
+        this.setState({ redirectSuccess: true, loading: false })
+      } else {
+        this.setState({ error: `Donation failed: ${response.message}`, loading: false })
+      }
+    } catch (e) {
+      this.setState({ error: e.message, loading: false })
+    }
+  }
+
+  render () {
+    const { redirectSuccess, loading } = this.state
+
+    if (redirectSuccess) {
+      Router.push('/ticket-success')
+      return (
+        <div />
+      )
+    }
+
+    return (
+      <Formik
+        initialValues={{
+          frequency: fundraisingEventTicket,
+          firstName: '',
+          lastName: '',
+          email: '',
+          amount: '100'
+        }}
+        validate={(values) => {
+          const { firstName, lastName, email } = values
+          const errors = {}
+          const firstNameError = validateText(firstName)
+          if (firstNameError) {
+            errors.firstName = firstNameError
+          }
+          const lastNameError = validateText(lastName)
+          if (lastNameError) {
+            errors.lastName = lastNameError
+          }
+          const emailError = validateEmail(email)
+          if (emailError) {
+            errors.email = emailError
+          }
+          return errors
+        }}
+        onSubmit={async (values, opts) => {
+          await this.donate(values)
+          opts.setSubmitting(false)
+        }}
+      >
+        {({
+          values,
+          errors,
+          touched,
+          handleChange,
+          handleBlur,
+          handleSubmit,
+          setFieldValue,
+          isSubmitting
+        }) => (
+          <Form className='flex flex-column f4-m ph2' onSubmit={handleSubmit}>
+            <div className='error tf-lato tc'>
+              <p className='red' aria-live='assertive'>{this.state.error}</p>
+            </div>
+
+            <DonationFrequency
+              name='frequency'
```

Comment:
I understand what's going on but entire file would make way more sense if you made this component more generic or copied it

Reply:
:)